asserts/database.go

```diff
@@ -742,43 +764,97 @@ type timestamped interface {
 		}
 	}
-	return nil
+	return delegationConstraints, nil
 }
 
-// XXX: keeping these in this form until we know better
-
 // A consistencyChecker performs further checks based on the full
 // assertion database knowledge and its own signing key.
 type consistencyChecker interface {
 	checkConsistency(roDB RODatabase, signingKey *AccountKey) error
 }
 
 // CheckCrossConsistency verifies that the assertion is consistent with the other statements in the database.
-func CheckCrossConsistency(assert Assertion, signingKey *AccountKey, roDB RODatabase, checkTimeEarliest, checkTimeLatest time.Time) error {
+func CheckCrossConsistency(assert Assertion, signingKey *AccountKey, delegationConstraints []*AssertionConstraints, roDB RODatabase, checkTimeEarliest, checkTimeLatest time.Time) ([]*AssertionConstraints, error) {
 	// see if the assertion requires further checks
 	if checker, ok := assert.(consistencyChecker); ok {
-		return checker.checkConsistency(roDB, signingKey)
+		return delegationConstraints, checker.checkConsistency(roDB, signingKey)
 	}
-	return nil
+	return delegationConstraints, nil
+}
+
+// CheckDelegationIsNotExpired checks that there is at least one not expired delegation constraint.
+func CheckDelegationIsNotExpired(assert Assertion, signingKey *AccountKey, delegationConstraints []*AssertionConstraints, roDB RODatabase, checkTimeEarliest, checkTimeLatest time.Time) ([]*AssertionConstraints, error) {
+	if len(delegationConstraints) == 0 {
+		// nothing to do
+		return nil, nil
+	}
+	notExpired := make([]*AssertionConstraints, 0, len(delegationConstraints))
+	for _, ac := range delegationConstraints {
+		if ac.isValidAssumingCurTimeWithin(checkTimeEarliest, checkTimeLatest) {
+			notExpired = append(notExpired, ac)
+		}
+	}
+	if len(notExpired) == 0 {
+		return nil, fmt.Errorf("all constraints supporting delegated %s assertion from %q to %q are expired", assert.Type().Name, assert.AuthorityID(), assert.SignatoryID())
+	}
+	return notExpired, nil
+}
+
+// CheckTimestampVsDelegationValidity verifies that the timestamp of
+// the assertion is within at least one of the delegation constraints validity.
+func CheckTimestampVsDelegationValidity(assert Assertion, signingKey *AccountKey, delegationConstraints []*AssertionConstraints, roDB RODatabase, checkTimeEarliest, checkTimeLatest time.Time) ([]*AssertionConstraints, error) {
+	if len(delegationConstraints) == 0 {
+		// nothing to do
+		return nil, nil
+	}
+	if tstamped, ok := assert.(timestamped); ok {
+		checkTime := tstamped.Timestamp()
+		valid := make([]*AssertionConstraints, 0, len(delegationConstraints))
+		for _, ac := range delegationConstraints {
+			if ac.isValidAt(checkTime) {
+				valid = append(valid, ac)
+			}
+		}
+		if len(valid) == 0 {
+			return nil, fmt.Errorf("delegated %s assertion from %q to %q timestamp %q is outside all supporting delegation constraints validity", assert.Type().Name, assert.AuthorityID(), assert.SignatoryID(), checkTime)
```

Comment:
/outside/outside of/ ?

Reply:
I think both work, but "outside of" was used in the other error